### PR TITLE
proposals: add "etcd" as "Incubating" CNCF project

### PR DESCRIPTION
**Pull request by gyuho**
ref. https://github.com/cncf/toc/issues/136.

/cc @erinboyd @philips @xiang90 @bgrant0607 @jpbetz @d-nishi

### CODE REVIEW

File: proposals/etcd.adoc [hunk 1]
```diff
@@ -0,0 +1,59 @@
+== etcd Project Proposal
+
+*Name of Project*: etcd
+
+*Description*:
+
+etcd is a consistent distributed key-value store, designed to hold small amounts of data that can fit entirely in memory (although etcd still writes to disk for durabilities) and mainly used as a separate coordination service for other distributed systems like Kubernetes. Typical etcd cluster is distributed over 3 to 5 nodes, for high availability, while it prioritizes consistency and partition tolerance. Which means, it provides on logical cluster view of many physical servers. So long as majority of cluster is up, etcd continues to work, even under machine failures. This redundancy provides fault tolerance.
```
[the remark this answers is not in the document]
Comment (quinton-hoole) [reply proposals/etcd.adoc]: nits: 
A typical etcd cluster...
a majority.

File: proposals/etcd.adoc [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+etcd server implements Raft consensus algorithm for data replication. Raft is a leader-based protocol. Data is replicated from leader to follower; follower forwards proposals to leader, and leader decides what to commit or not. Leader persists and replicates an entry, once it has been agreed by the quorum of cluster. The underlying storage layer for Raft log is write-ahead log (WAL). Committed entries are written out to disk, so they can be replayed on restart. etcd uses gRPC for transport layer. Client employs HTTP/2 Ping for server health checking, and implements automatic failover under faulty networks.
```
Comment (quinton-hoole) [inline proposals/etcd.adoc]: Perhaps worth noting explicitly that leader-based replication does not provide horizontal write scalability.  Write (and consistent read) throughput is limited to durable write throughput of a single leader node.

File: proposals/etcd.adoc [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+*Issue tracker*: https://github.com/coreos/etcd/issues
+
+*Initial committers*: https://github.com/philips[Brandon Phillips] and https://github.com/xiang90[Xiang Li] at CoreOS (see https://github.com/coreos/etcd/blob/master/MAINTAINERS)
+
+*Infrastructure requirements (CI / CNCF Cluster)*: N/A
+
+*External dependencies*: https://github.com/coreos/etcd/blob/master/bill-of-materials.json
+
+*Statement on alignment with CNCF mission*: etcd has enabled adoption of cloud native systems: etcd is "container packaged" by publishing container image for every release, and etcd is a critical component and most reliable storage implementation to many "dynamically managed" systems like Kubernetes.
```
Comment (quinton-hoole) [inline proposals/etcd.adoc]: nit: This paragraph could do with some word-smithing.  "has enabled adoption" and "most reliable" are strong claims, and unnecessary, IMO, to justify alignment with the CNCF's mission:

https://www.cncf.io/about/charter/

**Comment by philips** [comment]
overall looks reasonable. I agree with @quinton-hoole's nits.

**Comment by gyuho** [comment]
Address @quinton-hoole's feedback. Thanks!

**Comment by caniszczyk** [comment]
RFC @cncf/toc - will leave this open for a week or so for community feedback

**Comment by philips** [comment]
We need to add own, monitor, manage, and maintain discovery.etcd.io to this list as well. https://github.com/coreos/etcd/issues/9965#issuecomment-411834753

**Comment by gyuho** [comment]
@philips That makes sense, since we are donating etcd.io as well.

File: proposals/etcd.adoc [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+*License*: https://github.com/coreos/etcd/blob/master/LICENSE[Apache License 2.0]
+
+*Maturity Level*: Incubating
+
+*Source control repositories*: https://github.com/etcd-io (to be moved
+from https://github.com/coreos/etcd/issues/9965[August 27, 2018])
+
+*Issue tracker*: https://github.com/coreos/etcd/issues
+
+*Initial committers*: https://github.com/philips[Brandon Phillips] and
```
Comment (bgrant0607) [inline proposals/etcd.adoc]: If the full list of committers/maintainers is in https://github.com/coreos/etcd/blob/master/MAINTAINERS, then I suggest we remove the 2 specific individuals from here. There are active maintainers from at least Redhat (CoreOS), Amazon, and Google, correct?

File: proposals/etcd.adoc [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+https://github.com/xiang90[Xiang Li] at CoreOS.
+See https://github.com/coreos/etcd/blob/master/MAINTAINERS[MAINTAINERS].
+
+*Infrastructure requirements (CI / CNCF Cluster)*: N/A
+
+*External dependencies*: https://github.com/coreos/etcd/blob/master/bill-of-materials.json
+
+*Statement on alignment with CNCF mission*: etcd helps adopt cloud native
```
Comment (bgrant0607) [inline proposals/etcd.adoc]: Maybe @philips could help improve the wording here

File: proposals/etcd.adoc (hunk 4 above)
Comment (philips) [reply proposals/etcd.adoc]: I think there is a misunderstanding of what "Initial Committers" means. Does it mean the people who were initially on the project when it started or the people who will be on the project when donated to CNCF?

For context Xiang and I started the project 5 years ago so I assume that is why we are listed here.

File: proposals/etcd.adoc (hunk 5 above)
Comment (philips) [reply proposals/etcd.adoc]: @bgrant0607 how about this?

A consistent and partition tolerant datastore, like etcd, is a base dependency for many cloud native architectures. They hold on to critical cluster configuration and locks while providing guarantees against individual machine failure, network partitions, or data center power loss. In the literature and ecosystem etcd, or similar systems (e.g. chubby, zookeeper, etc), provide the persistence layer for applications like Kubernetes, CoreDNS, Vitess, Borg, Mesos, and countless others.

**Comment by caniszczyk** [comment]
Another question, are you considering making the https://github.com/coreos/etcd-operator part of this contribution?

File: proposals/etcd.adoc (hunk 4 above)
Comment (gyuho) [reply proposals/etcd.adoc]: @bgrant0607 @philips I took this from https://github.com/cncf/toc/blob/master/proposals/grpc.adoc, where initial "git" committers are listed as "Initial committers" (ref. https://github.com/grpc/grpc-java/graphs/contributors?from=2014-05-04&to=2014-09-30&type=c).

Comment (bgrant0607) [reply proposals/etcd.adoc]: Sorry, this may not be all that important, but perhaps @caniszczyk has guidance.

Comment (gyuho) [reply proposals/etcd.adoc]: @caniszczyk Shall we remove?

Comment (caniszczyk) [reply proposals/etcd.adoc]: @gyuho this means who has commit access to the project when donated, essentially we want to see who has write access, how diverse the committership is etc

Comment (gyuho) [reply proposals/etcd.adoc]: @caniszczyk @bgrant0607 Updated to only list maintainers who have write access to etcd repository. Look forward to expanding more!

File: proposals/etcd.adoc (hunk 5 above)
Comment (bgrant0607) [reply proposals/etcd.adoc]: LGTM, thanks

**Comment by bgrant0607** [comment]
@gyuho Is it correct that etcd has just one repo now? It looks like go-etcd, etcdctl, and etcd-ca repos are deprecated?

Is discovery.etcd.io still in use?

**Comment by gyuho** [comment]
> go-etcd, etcdctl, and etcd-ca repos are deprecated?

Yes. They were deprecated.

All core etcd components (package raft, "etcd" command, "etcdctl" command) are in one https://github.com/coreos/etcd repository.

> Is discovery.etcd.io still in use?

Some people still use it. @philips should have more data on this.

File: proposals/etcd.adoc (hunk 5 above)
Comment (gyuho) [reply proposals/etcd.adoc]: @philips @bgrant0607 Just updated this part. Thanks!

File: proposals/etcd.adoc [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+*Social media accounts*: None
+
+*Community size and any existing sponsorship*:
+
+Adopters: Alibaba, Amazon, Baidu, Cisco, CoreOS (now Red Hat), EMC, Google,
+Huawei, IBM, NTT, Uber, Verizon, Yandex, etc.
+See https://github.com/coreos/etcd/blob/master/Documentation/production-users.md[etcd production users] for more.
+
+Integrations: Kubernetes API server persists cluster metadata in etcd.
+OpenStack uses etcd to keep track service liveness. CockroachDB, TiKV, and
```
Comment (jpbetz) [inline proposals/etcd.adoc]: "keep track of"

Comment (gyuho) [reply proposals/etcd.adoc]: Fixed, thanks @jpbetz !

**Comment by philips** [comment]
@bgrant0607 discovery.etcd.io is still in extremely wide use and is used by default in OpenStack spin-up scripts. We had an outage recently on that service and dozens of people noticed on twitter, github, etc.

**Comment by jonboulle** [comment]
@caniszczyk where did this end up?

**Comment by caniszczyk** [comment]
@jonboulle voting is open: https://lists.cncf.io/g/cncf-toc/message/2237

The project team is still finalizing on when they would like to make the announcement, our approach has been to work with the project communities on their preferred time of announcement (there are enough votes for this to technically pass)

**Comment by caniszczyk** [comment]
+1 binding TOC votes (8/9):

Quinton: https://lists.cncf.io/g/cncf-toc/message/2259
Camille: https://lists.cncf.io/g/cncf-toc/message/2262
Brian: https://lists.cncf.io/g/cncf-toc/message/2268
Ben: https://lists.cncf.io/g/cncf-toc/message/2270
Sam: https://lists.cncf.io/g/cncf-toc/message/2271
Alexis: https://lists.cncf.io/g/cncf-toc/message/2278
Ken: https://lists.cncf.io/g/cncf-toc/message/2308
Jon: https://lists.cncf.io/g/cncf-toc/message/2431

Welcome etcd to the CNCF community :)

**Pull request merged**